### ISSUE TITLE
Segfault with specific ARM optimizations

### ISSUE THREAD

**Issue by fr500** [libretro/libretro-meta#19]
_From @zgoda on January 9, 2017 13:53_

Platform: Allwinner H3 (armv7-a, NEON, VFP4)

According to spec, this platform supports `-mfpu=neon-vfpv4` GCC flag. Compilation succeeds but resulting core segfaults. Removing FPU flag produces working core.

Failing optimization flags set: `-march=armv7-a -mfpu=neon-vfpv4 -mfloat-abi=hard -ftree-vectorize -funsafe-math-optimizations -pipe`, this is suggested for this platform by manufacturer.

Other libretro cores work fine with these flags, checked with snes9x-next, mupen64 and desmume.

_Copied from original issue: libretro/mgba#58_

**Comment by fr500** [comment]
_From @endrift on January 11, 2017 5:51_

I really wouldn't suggest `-funsafe-math-optimizations`. I don't believe it affects performance in mGBA anyway, but it can lead to, well, unsafe things happening.

Can you gather a backtrace from where it's crashing?

**Comment by fr500** [comment]
_From @zgoda on January 11, 2017 12:11_

Will try to find specific place where it dies.

I found NEON SIMD has something to do here. For Cortex-A7 cores recommended fpu flags are `neon-vfpv4`, `vfpv4` (without SIMD) or `vfpv4-d16` (without SIMD, 16 registers less) and only `neon-vfpv4` build segfaults. And this issue is related to this specific SOC, because exactly the same flags set work for eg. Raspberry Pi 2 (ARMv7-a, Cortex-A7) so I'm positive it's either bug in GCC or buggy SIMD implementation.

As for `unsafe-math-optimizations` flag, recommended `fast-math` flag sets it among other things on this platform so I guess it should be harmless but I will check this anyway.

I will try to produce any core dump for debugging.

**Comment by zgoda** [comment]
I got this:
```
retroarch: symbol lookup error: mgba_libretro.so: undefined symbol: _to16Bit
```
And `fast-math` does not have any impact on this.

EDIT: this comes from missing HAVE_NEON=1 in make invocation. With this appended core segfaults again.

**Comment by zgoda** [comment]
The only reference to this specific symbol I found is an old thread on [RetroPie github repo](https://github.com/RetroPie/RetroPie-Setup/issues/1029), they call this "Debian Jessie specific" but no details given. This is related to some extent because I'm building and running it on Debian Jessie based platform.

**Comment by zgoda** [comment]
Bit of digging and https://github.com/libretro/mgba/pull/18 sheds some light. Will test and report back.

**Comment by endrift** [comment]
It shouldn't be trying to use the _to16Bit function without NEON. See https://github.com/libretro/mgba/blob/998dc43314e6572c8e651d0f46afbefcf7d6aa47/src/gba/renderers/video-software.c#L596.

E] I think whoever added HAVE_NEON did not realize that the core still tries to use NEON if the compiler allows it, even without that one file, causing the ld issue.

**Comment by zgoda** [comment]
This was with `-mfpu=neon-vfpv4` in CFLAGS and without `HAVE_NEON=1` in `make` invocation.

Anyway by randomly shooting in the dark I got working set of CFLAGS that enable NEON optimizations on Allwinner H3:

```
-O3 -marm -march=armv7-a -mtune=cortex-a7 -mfpu=neon-vfpv4 -mfloat-abi=hard -funroll-loops -ffast-math
```

No matter what I tried, I could not get core file dumped. I'm not very satisfied with core performance so I will continue my investigations.

**Comment by endrift** [comment]
> This was with -mfpu=neon-vfpv4 in CFLAGS and without HAVE_NEON=1 in make invocation.

Yeah that's the problem. The person who added a flag for disabling NEON didn't add an ifdef for the code that uses it. It just excludes the file, without telling the code not to use that file.

**Comment by endrift** [comment]
cc @joolswills
